rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactory.groovy

```diff
@@ -16,27 +16,63 @@
 
 package com.netflix.spinnaker.rosco.providers.util
 
+import com.netflix.frigga.ami.AppVersion
 import com.netflix.spinnaker.rosco.api.BakeOptions
 import com.netflix.spinnaker.rosco.api.BakeRequest
+import com.netflix.spinnaker.rosco.api.BakeRequest.PackageType
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
 
-interface ImageNameFactory {
+import java.time.Clock
 
   /**
-   * This method is responsible for:
-   *   - Producing an image name.
-   *   - (If one or more package names are specified) Attempting to derive the AppVersion descriptor from the
-   *     first package name. If the AppVersion could not be derived, this value will be null.
-   *
-   * This method always returns a list of size 3 with the following elements:
-   *   1) A derived image name (to be used for naming the image being baked).
-   *   2) The appversion string (to be used for tagging the newly-baked image).
-   *   3) The updated list of packages to be used for overriding the passed package list (this will be removed once the
-   *      temporary workaround described above is removed).
-   *
-   * This function is not required to return the same image name on multiple invocations with the same bake request.
-   *
-   * Returns [imageName, appVersionStr, packagesParameter].
+   * Default implementation of ImageNameFactory considers only package_name, a timestamp, and base_os.
+   * For more fine grained conventions, extend this class, override the required methods and alter your
```

Comment:
Does this work as described here? Does each `CloudProviderBakeHandler` resolve the correct `@Bean` factory method from that provider's configuration class when autowiring the `ImageNameFactory`?
In any case, I think it would be easier to follow if `CloudProviderBakeHandler` defined an abstract `getImageNameFactory` method, instead of relying on the autowiring for this.